docs/content/doc/contributing/guidelines-backend.en-us.md

```diff
@@ -104,6 +104,46 @@ i.e. `services/user`, `models/repository`.
 Since there are some packages which use the same package name, it is possible that you find packages like `modules/user`, `models/user`, and `services/user`. When these packages are imported in one Go file, it's difficult to know which package we are using and if it's a variable name or an import name. So, we always recommend to use import aliases. To differ from package variables which are commonly in camelCase, just use **snake_case** for import aliases.
 i.e. `import user_service "code.gitea.io/gitea/services/user"`
 
+### Data Migrations
+
+Whenever you change the database structure, you **must** add a migration as well.
+Migrations are located under `models/migrations/v<next_gitea_version>/<version ID>.go`, and should be called in `models/migrations/migrations.go` using `NewMigration("<human readable title of what the migration does>", <migration function>)`
+
+#### Adding data
+
+If you only add data (structures), your migration function will look like the following example that actually exists:
+```go
+func AddVersionToActionRunner(x *xorm.Engine) error {
+	type ActionRunner struct {
+		Version string `xorm:"VARCHAR(64)"`
+	}
+
+	return x.Sync(new(ActionRunner))
+```
+As you can see, in this case you simply declare a function that takes `x *xorm.Engine` as parameter, and returns an `error` whether the migration was successful.
+Then, you declare a local migration type that shows what changes.
+Lastly, you call `return x.Sync(new(<migration type>))` to commit the migration and return the error it may produce.
+If your table didn't exist previously, it will be created.
+All columns present in your new type will also be created if they didn't exist yet.
+Existing columns that don't exist in your migration type are unaffected.
+If your migration type uses custom types, i.e. a new `int` type to simulate an enum, declare your migration type with `int` instead of `<your enum type>`. This helps to keep migrations consistent, even if the base type of your custom type is changed later.
+
+#### Deleting data
+
+In case you delete a column or table, please adapt the original migration that added it (and all subsequent migrations that use the deleted value) as well so that the column or table isn't added in the first place for new instances.
+Keep the original migration as a no-op (with a comment) in case it becomes empty.
```

Comment:
Is this what actually happens? I was always under the impression that past migrations should be left alone. There's less chance that an intermediate migration is missed that would cause some odd potential bug.

Reply:
Well… It's at least what I've observed with previous deletions, see for example #23605…
we can also decide to use another process, but we should decide this now.

Reply:
Hmm, tricky topic. I think leaving the original as-is may be safer than no-oping it. Treat it kind of like a git history that is never rewritten.

Reply:
In theory, I agree.
However, as you saw in #23605, it sometimes needs to be done.
What about a paragraph such as
> If you delete a column or table, do not adapt the original migration adding it and all migrations inbetween.
There is only one exception: If the original migration was broken and didn't work on some databases, you can convert the original migration and every change in between referencing the deleted content to a no-op.

?

Reply:
Yeah sounds good. General rule should be to not alter old migrations except in the rare case they are broken, in wich case it is warranted to fix it and "rewrite history".

Reply:
@delvh will you update with it?